mp4parse/src/lib.rs

```diff
@@ -55,24 +55,16 @@ pub fn vec_push<T>(vec: &mut Vec<T>, val: T) -> std::result::Result<(), ()> {
     Ok(())
 }
 
-#[allow(unreachable_code)]
-pub fn vec_reserve<T>(vec: &mut Vec<T>, size: usize) -> std::result::Result<(), ()> {
-    #[cfg(feature = "mp4parse_fallible")]
-    {
-        return FallibleVec::try_reserve(vec, size);
-    }
-
-    vec.reserve(size);
-    Ok(())
-}
-
 #[allow(unreachable_code)]
 fn allocate_read_buf(size: usize) -> std::result::Result<Vec<u8>, ()> {
     #[cfg(feature = "mp4parse_fallible")]
     {
         let mut buf: Vec<u8> = Vec::new();
         FallibleVec::try_reserve(&mut buf, size)?;
-        unsafe { buf.set_len(size); }
+        unsafe {
+            buf.set_len(size);
+            std::ptr::write_bytes(buf.as_mut_ptr(), 0, buf.len() * std::mem::size_of::<u8>());
+        }
```

Comment:
You can do `buf.extend(std::iter::repeat(0).take(size))` to avoid unsafe code. A loop on  `0..size` with `Vec::push` calls would probably also compile down to the same thing.

Reply:
Possibly done the way it is to avoid needing to reserve more space? Let me know if I'm mistaken, but if we `try_reserve(buf, n)`, `buf.set_len(n)`, then zero the vec with a ptr, we don't trigger further reserves (`set_len` guarantees it won't modify buffers). If we extend or push we get will get an extra allocation when we push at len == cap.

Reply:
This is incorrect, the `Vec` API won’t reserve additional space in the case the iteration fits, and it does here.

Reply:
Note that [`Vec::push`](https://github.com/rust-lang/rust/blob/74e35d270067afff72034312065c48e6d8cfba67/src/liballoc/vec.rs#L1019-L1030) itself will never reallocate as long as it has space, too. In general, don't use unsafe code if you don't need to.

Reply:
Thanks for the correction. I was mistaken as to when the new reserves take place (I was under the impression it was at the time len hit cap).